### PR TITLE
feat: Envelope serialization

### PR DESCRIPTION
**Pull request by bruno-garcia**

### CODE REVIEW

File: Sources/Sentry/SentryFileManager.m [hunk 1]
```diff
@@ -149,14 +149,10 @@ - (NSString *)storeData:(NSData *)data toPath:(NSString *)path {
...
-        return [self storeData:saveData toPath:path];
-    } else {
-        [SentryLog logWithMessage:[NSString stringWithFormat:@"Invalid JSON, failed to write file %@", path]
-                         andLevel:kSentryLogLevelError];
-    }
-    return path;
+    NSData *saveData = [SentrySerialization dataWithJSONObject:dictionary options:0 error:nil];
+    return nil != saveData
+            ? [self storeData:saveData toPath:path]
+            : path; // TODO: Should we return null instead? Whoever is using this return value is being tricked.
```
Comment (marandaneto) [inline Sources/Sentry/SentryFileManager.m]: is this gonna be addressed or can it be ignored for now?

File: Sources/Sentry/SentryNSURLRequest.m [hunk 2]
```diff
@@ -54,16 +54,17 @@ - (_Nullable instancetype)initStoreRequestWithDsn:(SentryDsn *)dsn
...
     } else {
         NSDictionary *serialized = [event serialize];
-        if (![NSJSONSerialization isValidJSONObject:serialized]) {
+        jsonData = [SentrySerialization dataWithJSONObject:serialized
+                                                   options:[SentrySDK.currentHub getClient].options.logLevel == kSentryLogLevelVerbose
+                                                           ? NSJSONWritingPrettyPrinted : 0
+                                                     error:error];
+        if (nil == jsonData) {
             if (error) {
+                // TODO: We're possibly overriding an error set by the actual code that failed ^
```
Comment (marandaneto) [inline Sources/Sentry/SentryNSURLRequest.m]: same as above comment

File: Sources/Sentry/SentrySerialization.m [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    // Parse items
+    NSInteger itemHeaderStart = envelopeHeaderIndex + 1;
+
+    NSMutableArray<SentryEnvelopeItem *> *items = [NSMutableArray new];
+    int endOfEnvelope = data.length - envelopeHeaderIndex;
+    for (NSInteger i = itemHeaderStart; i <= endOfEnvelope; ++i) {
+        if (bytes[i] == '\n' || i == endOfEnvelope) {
+            if (endOfEnvelope == i) i++; // 0 byte attachment
+            NSData *itemHeaderData = [data subdataWithRange:NSMakeRange(itemHeaderStart, i - itemHeaderStart)];
+#ifdef DEBUG
```
Comment (marandaneto) [inline Sources/Sentry/SentrySerialization.m]: should we not look for `options.debug` instead?

File: Sources/Sentry/SentryNSURLRequest.m (hunk 2 above)
Comment (bruno-garcia) [reply Sources/Sentry/SentryNSURLRequest.m]: this problem was already there, I'm just pointing it out.
I would rather address these separately on a phase before GA as we did for Android

File: Sources/Sentry/SentrySerialization.m (hunk 3 above)
Comment (bruno-garcia) [reply Sources/Sentry/SentrySerialization.m]: Not sure we need to print the whole envelope in prod builds if Options.Debug = true.
This was just while building the SDK itself in Debug (while writing the parsing code).

File: Sources/Sentry/SentryFileManager.m (hunk 1 above)
Comment (bruno-garcia) [reply Sources/Sentry/SentryFileManager.m]: This is the current behavior, I'm again just pointing out the issue.
Same as below

File: Sources/Sentry/SentrySerialization.m (hunk 3 above)
Comment (marandaneto) [reply Sources/Sentry/SentrySerialization.m]: yep, agreed, if its printing plain text, it won't make sense

**Pull request merged**